### DISCUSSION TITLE
Advanced initialization and getServerSession problem

### DISCUSSION THREAD

**Discussion by andreagiunti1** [nextauthjs/next-auth#6912] [Help]
Hello,

Using Advance initialization I created the options object to pass to NextAuth dynamically as explained in the documentation:

![1](https://user-images.githubusercontent.com/17339516/224338387-b22c2554-5fcc-4e2e-96fe-1be4a2ed3875.png)

In doing so, however, I can't find a system to export the object mentioned above to be able to import it and use it in getServerSession as indicated in the documentation:

![2](https://user-images.githubusercontent.com/17339516/224338423-1fb5d48e-6601-494b-8a08-140e2b1d047b.png)

From the documentation it seems that in order to use getServerSession, the AuthOptions object has to be created statically.
Is it possible to use getServerSession instead even if AuthOptions is created dynamically when a request is made to /api/auth/[...nextauth]?
Thanks in advance.

**Comment by pbell23** [comment]
Any update on this ?
I also don't see how to recover the session when using advanced initialization. it's rather blocking

**Reply by andreagiunti1** [reply]
The only solution i've found is to use getSession instead of getServerSession, but in the documentation they said to not use it... I hope someone will answer with a better solution.

**Comment by ekofi** [comment]
Any update?

**Comment by jusiho** [comment]
find an answer?

**Comment by rexfordessilfie** [comment]
I ran exactly into this recently and I'm happy to share here! How you can approach this is the following:
1. Define `authOptions` as a function instead, for example: `getNextAuthOptions`, that takes a `req` and a `res` object
2. And then instead of passing in `authOptions` to `getServerSession`, you instead pass in a call to the function `getNextAuthOptions(req, res)`
3. This should work since the call to `getServerSession` requires the `req` and `res` objects as well (unless the API changes in the future)!

Happy to share a concrete example in an actual codebase if it helps!

**Reply by rexfordessilfie** [reply]
Hi, @balazsorban44! Please do you think it is worth adding this to the NextAuth docs? I have encountered two questions on this so far and ran into this use case as well:
* https://github.com/nextauthjs/next-auth/discussions/6912#discussioncomment-6025442
* https://github.com/nextauthjs/next-auth/discussions/5069#discussion-4272912

I'm happy to make the change and open a PR to do so and with a minimal code example!

**Reply by andreagiunti1** [reply]
Thanks a lot man! If you could share an example it would be amazing!

**Reply by rexfordessilfie** [reply]
@andreagiunti1, sure thing! Here you go:


1. `pages/api/auth/[...nextauth].ts`
```typescript
// ...
import {
  GetServerSidePropsContext,
  NextApiRequest,
  NextApiResponse,
} from "next";
import NextAuth, { NextAuthOptions } from "next-auth";

export const getNextAuthOptions = (
  req: NextApiRequest | GetServerSidePropsContext["req"],
  res: NextApiResponse | GetServerSidePropsContext["res"]
) => {
  // Accessing the `req` and `res` objects within next auth options
  console.log(res.getHeaders());
  console.log(req.headers);

  const authOptions: NextAuthOptions = {
    providers: [
      // ...
    ],
    callbacks: {
        // ...
    },
  };

  return authOptions;
};

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  return NextAuth(req, res, getNextAuthOptions(req, res));
}
```

2. `pages/index.tsx`
```typescript
//...

import { getNextAuthOptions } from "./api/auth/[...nextauth]";

export default function Page(){
  // ...
  return (...)
}

export async function getServerSideProps(context: GetServerSidePropsContext) {
  const session = await getServerSession(
    context.req,
    context.res,
    getNextAuthOptions(context.req, context.res)
  );
  
  return {
    props: {
      session,
    },
  };
}
```

**Reply by HamedBahram** [reply]
Thanks for sharing this, Rexford!

**Reply by rexfordessilfie** [reply]
You’re welcome!

**Reply by leeren** [reply]
Any idea how to do this with Next.js 13?

**Comment by quan612** [comment]
Hi @andreagiunti1 

May I know how do you pass the req.body from front-end so that in ...nextauth you get access to it like your use-case:
`req.body.keepLogin`

is that through the Rest Api mentioned in next-auth document that we have to called manually instead of using signIn() built in.

**Reply by andreagiunti1** [reply]
Hi!
You can pass something with req.body with the signIn() method like this:
![image](https://github.com/nextauthjs/next-auth/assets/17339516/6253820c-d260-41c9-9668-58426fdf47b4)
![image](https://github.com/nextauthjs/next-auth/assets/17339516/7dc2b1a6-1fb8-4e41-a6bd-a40481f55e1b)
Then in nextauth file you can get it by using req.body.keepLoggedIn. 
I hope this helps you!
